src/bip324.cpp

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2023 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <bip324.h>
+
+#include <chainparams.h>
+#include <crypto/chacha20.h>
+#include <crypto/chacha20poly1305.h>
+#include <crypto/hkdf_sha256_32.h>
+#include <random.h>
+#include <span.h>
+#include <support/cleanse.h>
+
+#include <algorithm>
+#include <assert.h>
+#include <cstdint>
+#include <cstddef>
+
+BIP324Cipher::BIP324Cipher() noexcept
+{
+    m_key.MakeNewKey(true);
+    uint256 entropy = GetRandHash();
+    m_our_pubkey = m_key.EllSwiftCreate(MakeByteSpan(entropy));
+}
+
+BIP324Cipher::BIP324Cipher(const CKey& key, Span<const std::byte> ent32) noexcept :
+    m_key(key)
+{
+    m_our_pubkey = m_key.EllSwiftCreate(ent32);
+}
+
+BIP324Cipher::BIP324Cipher(const CKey& key, const EllSwiftPubKey& pubkey) noexcept :
+    m_key(key), m_our_pubkey(pubkey) {}
+
+void BIP324Cipher::Initialize(const EllSwiftPubKey& their_pubkey, bool initiator, bool self_decrypt) noexcept
+{
+    // Determine salt (fixed string + network magic bytes)
+    const auto& message_header = Params().MessageStart();
+    std::string salt = std::string{"dash_v2_shared_secret"} + std::string(std::begin(message_header), std::end(message_header));
```

Comment:
Do we really need to deviate from bip324 here? Isn't having different network magic bytes enough already? I _think_ we could simply go with the original salt, drop 566ba2a and it should just work.

Reply:
The salt is _composed_ of a fixed message and network bytes. Changing network bytes will therefore change the salt. We could reuse Bitcoin's salt but that would involve hardcoding their network bytes. I don't really see a good reason to do that.

Since we're already changing the salt due to the difference in network bytes, I went ahead and changed the fixed message too... either ways, without hardcoding Bitcoin's network magic, we'd have to regenerate values. This has been done before, the message magic for instance, [was changed for Dash](https://github.com/dashpay/dash/blob/5f6ae3ca14f5e68c77203fe946e90b0e51ad9e45/src/util/message.cpp#L18-L22) (with its former name).

Reply:
ah, right... will have to update test vectors anyway... hmm.

Reply:
>This has been done before, the message magic for instance, [was changed for Dash](https://github.com/dashpay/dash/blob/5f6ae3ca14f5e68c77203fe946e90b0e51ad9e45/src/util/message.cpp#L18-L22) (with its former name).

Yes, I remember that and that's exactly why I wanted to avoid making the same mistake here :D

Reply:
I'm leaning towards following bip324 and simply using our bytes. @PastaPastaPasta ?

Reply:
I *don't think* I agree with Udjin here; I am open to further discussion

The BIP even outlines the salt as something parameterized; I'm receptive to an argument that this may make us non-compatible with bip-324; I guess we *should* create a dip which specifies the modified salt? 
```
def initialize_v2_transport(peer, ecdh_secret, initiating):
    # Include NETWORK_MAGIC to ensure a connection between nodes on different networks will immediately fail
    prk = HKDF_Extract(Hash=sha256, salt=b'bitcoin_v2_shared_secret' + NETWORK_MAGIC, ikm=ecdh_secret)
```

Reply:
> The BIP even outlines the salt as something parameterized
> ...
> 
> ```
> def initialize_v2_transport(peer, ecdh_secret, initiating):
>     # Include NETWORK_MAGIC to ensure a connection between nodes on different networks will immediately fail
>     prk = HKDF_Extract(Hash=sha256, salt=b'bitcoin_v2_shared_secret' + NETWORK_MAGIC, ikm=ecdh_secret)
> ```

It explicitly states that `NETWORK_MAGIC` is what does the job here, not the (`bitcoin_v2_shared_secret`) prefix. Also, creating a DIP only to say that our version uses another string feels weird... like, it's not such a huge improvement to have its own DIP imo :D

Reply:
Hmm; @kwvg any input here? any reason we should change the static (non-network bytes component) shared secret? If having different network bytes is enough then...? why change this too?

Reply:
An earlier version of this PR changed didn't change the fixed message. The values were recomputed solely due to the difference in network bytes.

On remembering the change of  `MESSAGE_MAGIC` and taking it as precedent, the fixed message was changed and values recomputed again.

I don't think simply changing the fixed message would require issuing a DIP. It seems unusual for any (unlikely) third party implementation that supports Bitcoin forks and BIP324 to expect that the fixed message would be the same no matter what.

Retaining the Bitcoin branding for something like PSBT makes sense since the benefit of renaming it is nonexistent compared to the trouble it'd give to developers. Retaining it here in the fixed message doesn't cause harm or benefit. 

I didn't know that changing `MESSAGE_MAGIC` was regrettable. It does lock you into a value that will remain the same regardless of future rebranding but I lean towards the belief a magic value should be unique (not just unique enough).

But I don't feel strongly on the matter.

Reply:
@UdjinM6, the fixed message has been changed back to `bitcoin_v2_shared_secret` in the latest push ([diff](https://github.com/dashpay/dash/compare/9aee22f2067bc9f103d582e3ab8c6dac4a182c8f..da40c7da5fc7624e99782dd6f18cb84deb0f6927)).